api/src/task/notifications/closing_date_notification.py

```diff
@@ -0,0 +1,171 @@
+
+from src.adapters import db
+from src.db.models.opportunity_models import Opportunity, OpportunitySummary
+from src.db.models.user_models import UserOpportunityNotificationLog, UserSavedOpportunity
+from src.task.notifications.BaseNotification import BaseNotification
+from src.task.notifications.constants import EmailData, NotificationConstants
+from src.util import datetime_util
+
+logger = logging.getLogger(__name__)
+
+CONTACT_INFO = (
+    "mailto:support@grants.gov\n"
+    "1-800-518-4726\n"
+    "24 hours a day, 7 days a week\n"
+    "Closed on federal holidays"
+)
+
+
+class ClosingDateNotification(BaseNotification):
+    collected_data: dict | None = None
+
+    def __init__(self, db_session: db.Session):
+        super().__init__(
+            db_session,
+        )
+        self.frontend_base_url = None
+
+    def collect_notifications(self) -> dict[UUID, list[UserSavedOpportunity]] | None:
+        """Collect notifications for opportunities closing in two weeks"""
+        two_weeks_from_now = datetime_util.utcnow() + timedelta(days=14)
+
+        # Find saved opportunities closing in two weeks that haven't been notified
+        stmt = (
+            select(UserSavedOpportunity)
+            .join(UserSavedOpportunity.opportunity)
+            .join(
+                OpportunitySummary, OpportunitySummary.opportunity_id == Opportunity.opportunity_id
+            )
+            .where(
+                # Check if closing date is within 24 hours of two weeks from now
+                and_(
+                    OpportunitySummary.close_date >= two_weeks_from_now - timedelta(hours=24),
+                    OpportunitySummary.close_date <= two_weeks_from_now + timedelta(hours=24),
+                ),
+                # Ensure we haven't already sent a closing reminder
+                ~exists().where(
+                    and_(
+                        UserOpportunityNotificationLog.user_id == UserSavedOpportunity.user_id,
+                        UserOpportunityNotificationLog.opportunity_id
+                        == UserSavedOpportunity.opportunity_id,
+                    )
+                ),
+            )
+        )
+
+        results = self.db_session.execute(stmt).scalars().all()
+        closing_date_opportunities: dict[UUID, list[UserSavedOpportunity]] = {}
+
+        for result in results:
+            user_id = result.user_id
+            closing_date_opportunities.setdefault(user_id, []).append(result)
+
+        logger.info(
+            "Collected closing date notifications",
+            extra={
+                "user_count": len(closing_date_opportunities),
+                "total_notifications": sum(
+                    len(closing_date_opportunities)
+                    for container in closing_date_opportunities.values()
+                ),
+            },
+        )
+        self.collected_data = closing_date_opportunities or None
+
+        return closing_date_opportunities or None
```

Comment:
It generally makes logic a bit easier when working with lists/sets/collections, just return an empty list/set/whatever.

Normally we'll have logic like:

```py
my_values = get_some_list(...)
for value in my_values:
    print(value)
```
If `get_some_list` always returns a list, this works, if we can return None, then we have to additionally add guardrails if it is None. Better to just return an empty list. If an empty list is a sign of a problem, well then we can add a `if len(my_values) == 0`.

Not having anything to iterate over is often fine, keep it simple.